### PR TITLE
Don't load datepicker module until it is actually needed

### LINKED ISSUE #28823: [Issue] Don't load datepicker module until it is actually needed
###### (i) This ticket is created automatically based on a verified GitHub issue.
{panel}
***If you are starting to work on this Jira ticket, please notify issue reporter and authors of related community pull requests (if any) that you are going to work on this issue:***
 1. *Assign [#28823](https://github.com/magento/magento2/issues/28823) issue to yourself on Magento GitHub*
 2. *Add comment "Internal Magento team is working on this issue right now."*
{panel}
This issue is automatically created based on existing pull request: magento/magento2#27860: Don't load datepicker module until it is actually needed ~~-------~~ ###### Description (*)

Magento 2.4-develop
 This PR is a part of my effort to improve Magento 2 frontend performance.

Goal here is to eliminate unnecessary loading of datepicker UI component, which is only needed when the shop has date of birth field enabled.
###### Expected behavior (*)

1. Enable date of birth field for customers in admin.
 2. Verify that `jquery/ui-modules/datepicker.js` is not loaded on pages that do not use datepicker (e.g. homepage, category, product etc.).
 3. Verify that above file is loaded and datepicker works properly on pages that use it (e.g. customer registration form page).
###### Benefits

***Performance improvement***
 Tested on Magento 2.4-develop with Luma and sample data.
 - 205 -> 202 JS requests (-1.46%).
 - 698.40kB -> 652.57kB JS transferred (-6.56%).

###### Contribution checklist (*)
 - < ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ > All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by krzksz**
### Description (*)
This PR is a part of my effort to improve Magento 2 frontend performance.

Goal here is to eliminate unnecessary loading of datepicker UI component, which is only needed when the shop has date of birth field enabled.

#### Performance improvement

Tested on Magento 2.4-develop with Luma and sample data. 

- 205 → 202 JS requests (-1.46%).
- 698.40kB → 652.57kB JS transferred (-6.56%).

### Manual testing scenarios (*)

1. Enable date of birth field for customers in admin.
2. Verify that `jquery/ui-modules/datepicker.js` is not loaded on pages that do not use datepicker (e.g. homepage, category, product etc.).
3. Verify that above file is loaded and datepicker works properly on pages that use it (e.g. customer registration form page).

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#28823: Don't load datepicker module until it is actually needed

### CODE REVIEW

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @krzksz,
Please review failing tests, looks like date format has changed

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @krzksz,
Functional tests still fails due to incorrect date format, please fix it

**Comment by krzksz** [comment]
@magento give me test instance

**Comment by magento-engcom-team** [comment]
Hi @krzksz. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @krzksz, here is your [new Magento instance](https://pr-27860.instances.magento-community.engineering).
Admin access: [https://pr-27860.instances.magento-community.engineering/admin_be04](https://pr-27860.instances.magento-community.engineering/admin_be04)
Login: `de96da0c` Password: `f341cfebc4ef`
Instance will be terminated in up to 3 hours.

**Review by ihor-sviziev — APPROVED** [review]
✅ approved 
Failing functional tests for EE isn’t related to changes from this PR

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7459](https://jira.corp.magento.com/browse/ENGCOM-7459) has been created to process this Pull Request

**Comment by krzksz** [comment]
@ihor-sviziev I'm not sure about `Auto-Tests: Not Required` because I had to adjust unit tests in order for the new logic to pass them 🤔

**Comment by engcom-Alfa** [comment]
Hi @krzksz .
During testing, we faced with one question.

`Magento_Ui/js/lib/knockout/bindings/datepicker.js` loaded on pages that do not use datepicker (e.g. homepage, category, product).

Tested on Magento 2.4-develop with Luma and sample data.

**Actual Result:**
Peek 2020-04-21 12-29

@krzksz Could you take a look?
Thanks!

**Comment by krzksz** [comment]
@engcom-Alfa Sorry, it was my mistake in the description. The file that should not be loaded is actually `jquery/ui-modules/datepicker.js`, while this binding should be loaded always just as on your screenshot.

**Comment by engcom-Alfa** [comment]
@krzksz Thank you for the clarification!

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

Tested on Magento 2.4-develop with Luma and sample data.

**Manual testing scenario:**
1. Enable date of the birth field for customers in admin;
2. Go to Storefront (Home page) and open developer tool

**Before:** :heavy_multiplication_x:` jquery/ui-modules/datepicker.js` loaded on pages that do not use datepicker (e.g. homepage, category, product etc.).
after

**After:** :heavy_check_mark: ` jquery/ui-modules/datepicker.js` isn't loaded on pages that do not use datepicker (e.g. homepage, category, product etc.)
Screenshot from 2020-04-21 12-54-27

3. Go to the Customer registration page;

**Actual Result:**
:heavy_check_mark: `Magento_Ui/js/lib/knockout/bindings/datepicker.js` is loaded and datepicker works properly (tested on _customer registration_ page and _edit customer information page_)
Screenshot from 2020-04-21 13-10-19

**Performance improvement:**
212 -> 209 requests (-1.46%)
652kB → 606kB JS transferred (-7.5%).

**Review by ptylek — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ptylek, thank you for the review.
[ENGCOM-7459](https://jira.corp.magento.com/browse/ENGCOM-7459) has been created to process this Pull Request

File: app/code/Magento/Ui/view/base/web/js/lib/knockout/bindings/datepicker.js [hunk 1]
```diff
@@ -75,26 +75,21 @@ define([
                 observable = config;
             }
 
-            if (_.isEmpty(observable())) {
-                if ($(element).datepicker('getDate')) {
-                    $(element).datepicker('setDate', null);
-                    $(element).blur();
+            require(['moment', 'mage/utils/misc', 'mage/calendar'], function (moment, utils) {
```
Comment (DrewML) [inline app/code/Magento/Ui/view/base/web/js/lib/knockout/bindings/datepicker.js:78]: Looks like, if we move the `_.isEmpty(observable)` check outside the `require`, we can avoid loading `moment` and `mage/utils/misc` in some cases, right?

**Review by DrewML — APPROVED** [review]
Fantastic work! Suggested one additional optimization, but looks great!

**Comment by engcom-Delta** [comment]
@magento create issue

**Pull request merged**